### PR TITLE
Add more prometheus metrics

### PR DESCRIPTION
**Pull request by jereksel**
This MR implement more metrics mentioned [here](https://github.com/containous/traefik/issues/156#issuecomment-221536251).

It implements more detailed backend metrics (it now includes backend server), frontend metrics (see below) and config reload metrics.

Unfortunatelly there doesn't seem to be a clean way to get more information about frontend without rewriting route management.

I'm also not sure about frontend/backend times. I've included debug start/stop messages for backend and frontend and these are the results:

```
backend start
backend start
frontend start
DEBU[2017-04-09T21:03:27+02:00] Round trip: http://httpbin.org/, code: 200, duration: 165.885424ms 
frontend stop
backend stop
backend stop
```
Which is very weird because backend is invoked before frontend and I don't know how to add time metrics that will correctly calculate frontend and backend response times.

### CODE REVIEW

**Comment by timoreimann** [comment]
@jereksel you cannot change vendored code directly; you have to make a PR against upstream and afterwards revendor the updated library in Traefik (i.e., use glide to bump our referenced version).

For the time being though, I think it's okay to pretend this is the oxy change we want, do the review, and eventually update containous/oxy (our fork of vulcand/oxy) should we all agree to the suggested path.

File: server.go [hunk 1]
```diff
@@ -219,17 +230,22 @@ func (server *Server) listenProviders(stop chan bool) {
...
 				return
 			}
+			var labels []string
 			server.defaultConfigurationValues(configMsg.Configuration)
 			currentConfigurations := server.currentConfigurations.Get().(configs)
 			jsonConf, _ := json.Marshal(configMsg.Configuration)
 			log.Debugf("Configuration received from provider %s: %s", configMsg.ProviderName, string(jsonConf))
 			if configMsg.Configuration == nil || configMsg.Configuration.Backends == nil && configMsg.Configuration.Frontends == nil {
 				log.Infof("Skipping empty Configuration for provider %s", configMsg.ProviderName)
+				labels = []string{"state", "failed", "type", configMsg.ProviderName}
```
Comment (timoreimann) [inline server.go]: The only parameter that changes between the different occasions where we assign the slice is the state. Can we initialize the slice with the other parameters in line 236 right away and just complete the state in the subsequent lines, respectively?

File: middlewares/transformer.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package middlewares
+
+import (
+	"net/http"
+)
+
+//Transformer is a hacky way to get metrics middleware with frontend
```
Comment (timoreimann) [inline middlewares/transformer.go]: Not sure I get why we need the transformer. Could you explain please?

File: middlewares/metrics.go [hunk 3]
```diff
@@ -19,13 +24,41 @@ type Metrics interface {
 // given Metrics implementation to expose and monitor Traefik metrics
 type MetricsWrapper struct {
 	Impl Metrics
+
+	//frontend/backend
```
Comment (timoreimann) [inline middlewares/metrics.go]: The comment should maybe be a bit more elaborate.

Also, space between slash and first word.

**Comment by ldez** [comment]
@jereksel do you the time to update this PR before the upcoming feature freeze for the 1.3 release?

File: middlewares/metrics.go [hunk 4]
```diff
@@ -34,9 +67,21 @@ func NewMetricsWrapper(impl Metrics) *MetricsWrapper {
...
+	var state string
+	if prw.StatusCode() < 400 {
+		state = "Successful"
+	} else {
+		state = "Failing"
+	}
+
+	labelsStatus := []string{"state", state, "method", r.Method, "name", m.getName(r), "type", m.typ}
 	m.Impl.getReqsCounter().With(labels...).Add(1)
+	m.Impl.getReqsStatusCounter().With(labelsStatus...).Add(1)
```
Comment (jereksel) [inline middlewares/metrics.go]: Sure, I was basing it on rancher-exporter that splits these metrics. But having them together is indeed better idea.

File: middlewares/metrics.go [hunk 5]
```diff
@@ -19,13 +24,41 @@ type Metrics interface {
 // given Metrics implementation to expose and monitor Traefik metrics
 type MetricsWrapper struct {
 	Impl Metrics
+
+	//frontend/backend
+	typ     string
```
Comment (jereksel) [inline middlewares/metrics.go]: I thought about naming it `type`, but it's reserved keyword. Phase sound good thought.

**Comment by jereksel** [comment]
Sure, I'll try to fix this MR today.

File: middlewares/transformer.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	next    http.Handler
+	metrics *MetricsWrapper
+}
+
+//NewTransformer creates new proxy for given Metrics and Handler
+func NewTransformer(next http.Handler, metrics *MetricsWrapper) *Transformer {
+	return &Transformer{next, metrics}
+}
+
+func (sb *Transformer) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
```
Comment (jereksel) [inline middlewares/transformer.go]: I copied it from another file :smile:

**Comment by jereksel** [comment]
@Idez @timoreimann Maybe I'll scrap this and redo it with https://github.com/containous/traefik/pull/1408/ and https://github.com/containous/traefik/pull/1485/

**Comment by ldez** [comment]
@jereksel simply rebase your branch, like that you'll already have the content of the #1408

**Comment by timoreimann** [comment]
Yep, rebase is the way to go.

We have also restructured/moved some packages. (For instance, `/server.go` now lives in `/server/server.go`.) My git client got a bit confused by that and tried to merge files into `integration`. I managed to solve this by specifying the rename threshold like this: `git rebase -X find-renames=25 upstream/master`.

**Comment by jereksel** [comment]
I think I'll wait for #1485 - I think I have an idea how to combine it with Prometheus metrics.

**Pull request closed without merging**